### ISSUE TITLE
MMC is broken on master

### ISSUE THREAD

**Issue by redknightlois** [NervanaSystems/coach#369]
Latest code raises the following error. Same code on 12.1 works properly.

```
Original stack trace for 'main_level/agent/main/online/network_0/behavior/behavior':
  File "C:\Anaconda3\lib\site-packages\rl_coach\graph_managers\graph_manager.py", line 148, in create_graph
    self.level_managers, self.environments = self._create_graph(task_parameters)
  File "C:\Anaconda3\lib\site-packages\rl_coach\graph_managers\basic_rl_graph_manager.py", line 72, in _create_graph
    level_manager = LevelManager(agents=agent, environment=env, name="main_level")
  File "C:\Anaconda3\lib\site-packages\rl_coach\level_manager.py", line 91, in __init__
    self.build(spaces_definition)
  File "C:\Anaconda3\lib\site-packages\rl_coach\level_manager.py", line 165, in build
    [agent.set_environment_parameters(spaces) for agent in self.agents.values()]
  File "C:\Anaconda3\lib\site-packages\rl_coach\level_manager.py", line 165, in <listcomp>
    [agent.set_environment_parameters(spaces) for agent in self.agents.values()]
  File "C:\Anaconda3\lib\site-packages\rl_coach\agents\agent.py", line 335, in set_environment_parameters
    self.init_environment_dependent_modules()
  File "C:\Anaconda3\lib\site-packages\rl_coach\agents\value_optimization_agent.py", line 46, in init_environment_dependent_modules
    super().init_environment_dependent_modules()
  File "C:\Anaconda3\lib\site-packages\rl_coach\agents\agent.py", line 380, in init_environment_dependent_modules
    self.networks = self.create_networks()
  File "C:\Anaconda3\lib\site-packages\rl_coach\agents\agent.py", line 353, in create_networks
    worker_device=self.worker_device)
  File "C:\Anaconda3\lib\site-packages\rl_coach\architectures\network_wrapper.py", line 86, in __init__
    network_is_trainable=True)
  File "C:\Anaconda3\lib\site-packages\rl_coach\architectures\tensorflow_components\general_network.py", line 74, in construct
    return construct_on_device()
  File "C:\Anaconda3\lib\site-packages\rl_coach\architectures\tensorflow_components\general_network.py", line 59, in construct_on_device
    return GeneralTensorFlowNetwork(*args, **kwargs)
  File "C:\Anaconda3\lib\site-packages\rl_coach\architectures\tensorflow_components\general_network.py", line 126, in __init__
    network_is_local, network_is_trainable)
  File "C:\Anaconda3\lib\site-packages\rl_coach\architectures\tensorflow_components\architecture.py", line 105, in __init__
    self.weights = self.get_model()
  File "C:\Anaconda3\lib\site-packages\rl_coach\architectures\tensorflow_components\general_network.py", line 262, in get_model
    input_placeholder, embedding = input_embedder()
  File "C:\Anaconda3\lib\site-packages\rl_coach\architectures\tensorflow_components\embedders\embedder.py", line 86, in __call__
    self.input = tf.placeholder("float", shape=[None] + self.input_size, name=self.get_name())
  File "C:\Anaconda3\lib\site-packages\tensorflow\python\ops\array_ops.py", line 2143, in placeholder
    return gen_array_ops.placeholder(dtype=dtype, shape=shape, name=name)
  File "C:\Anaconda3\lib\site-packages\tensorflow\python\ops\gen_array_ops.py", line 7400, in placeholder
    "Placeholder", dtype=dtype, shape=shape, name=name)
  File "C:\Anaconda3\lib\site-packages\tensorflow\python\framework\op_def_library.py", line 788, in _apply_op_helper
    op_def=op_def)
  File "C:\Anaconda3\lib\site-packages\tensorflow\python\util\deprecation.py", line 507, in new_func
    return func(*args, **kwargs)
  File "C:\Anaconda3\lib\site-packages\tensorflow\python\framework\ops.py", line 3616, in create_op
    op_def=op_def)
  File "C:\Anaconda3\lib\site-packages\tensorflow\python\framework\ops.py", line 2005, in __init__
    self._traceback = tf_stack.extract_stack()
```

**Comment by galnov** [comment]
The Doom_Health_MMC preset is running fine on my setup with the latest code on master, so need some more details:
1. Command line (or code) you're using to launch the training process 
2. Are you getting the failure immediately when starting to train?
3. Which Tensorflow version are you using?

**Comment by redknightlois** [comment]
1. I am launching using my own code cause I have a very complex data generation setup
2. On Heatup
3. Tensorflow 1.14

It also affects Rainbow and a few others it seems.

**Comment by redknightlois** [comment]
Code is pretty straightforward.

Sample:

```
agent_params = MixedMonteCarloAgentParameters()

agent_params.network_wrappers['main'].replace_mse_with_huber_loss = with_default(training_profile.replace_mse_with_huber_loss, True)

# agent_params.exploration.evaluation_epsilon = with_default(mmc_config.evaluation_epsilon, 0.0)
agent_params.memory.max_size = (MemoryGranularity.Episodes, with_default(training_profile.max_episodes, 200))


for key, embedder in agent_params.network_wrappers['main'].input_embedders_parameters.items():
	embedder.activation_function = 'leaky_relu'
	embedder.batchnorm = with_default(model_profile.batch_normalization, True)

graph_manager = BasicRLGraphManager(
        agent_params=agent_params,
        env_params=env_params,
        vis_params=VisualizationParameters(print_networks_summary=True, tensorboard=False),
        schedule_params=SimpleSchedule(),
    )

# Ensure initialization
params = TaskParameters(experiment_path=agents_dir,
                        checkpoint_save_dir=os.path.join(agents_dir, 'checkpoint'))

graph_manager.create_graph(params)
graph_manager.task_parameters.experiment_path = agents_dir
graph_manager.task_parameters.checkpoint_restore_path = os.path.join(agents_dir, 'checkpoint')

graph_manager.improve_steps.num_steps = num_steps
graph_manager.steps_between_evaluation_periods.num_steps = 500
iterations = int(with_default(training_profile.timesteps, 1000000) / num_steps)

episodes_to_heatup = with_default(training_profile.exploration_heatup, 100)
graph_manager.heatup(EnvironmentSteps(episodes_to_heatup * env.max_steps))

for i in range(0, iterations):
	printout(f'Executing {i+1}/{iterations} iterations')
	graph_manager.improve()
	graph_manager.save_checkpoint()
```

**Comment by galnov** [comment]
I suggest you try running the Doom_Health_MMC preset to see if this might be a system configuration problem. This preset is running fine for me with the latest master code and with Tensorflow 1.14.

If it is working fine, then the only change that might be related to this failure is the recent [batchnorm fix](https://github.com/NervanaSystems/coach/commit/d6795bd5245ba18604ce1b1ffa9fbee9c8ce7778), so I would check whether your code works before and after this change.

**Comment by redknightlois** [comment]
Easy enough. I can confirm that d6795bd5245ba18604ce1b1ffa9fbee9c8ce7778 doesn't work and the commit before does.

**Comment by gal-leibovich** [comment]
Hi @redknightlois,  d6795bd fixed an issue with batchnorm for DDPG. Unfortunately, even with DDPG, we did not see any significant improvements with batchnorm being used.

At the moment, batchnorm is not supported generically enough to be applied for any other algorithm, other than DDPG. We might want to fix that at some later point, but have no short term plans to support batchnorm for other algorithms.

**Comment by redknightlois** [comment]
@gal-leibovich Good to know that there is a way to avoid the issue, wanted to try TD3 but master was misbehaving. One thing that striked me as odd when reading the code was that coach applies BatchNorm, the linear + activation and then dropout.

Recent research has been showing that with batch normalization order matters. I don't have the reference fresh though, but there are a few papers about it in the wild. The guys at FastAI which had tried almost everything has been using BatchNorm, Dropout, Linear + Activation instead. 
Have you tried https://github.com/fastai/fastai/blob/master/fastai/layers.py#L44 ... I had been having excellent results on that setup on supervised environments.

**Comment by redknightlois** [comment]
@gal-leibovich This paper may shed some light into why BatchNorm is not showing improvements and also suggest a different normalization technique that may be of interest. https://arxiv.org/pdf/1902.05605.pdf
